docs/design-documents/platform/clocks/clocks.md

```diff
@@ -0,0 +1,418 @@
+### Using the high-resolution clock
+
+**Measuring elapsed time using Timer**
+```C++
+template<class F>
+microseconds time_operation(const F &operation)
+{
+    Timer timer;
+    timer.start(); // deep sleep is automatically locked while timer is running
+    operation();
+    return timer.elapsed_time();
+} // Deep sleep automatically unlocked when timer is destroyed.
+
+microseconds print_time = time_operation([] { printf("Hello!\n"); });
+printf("Printing that took %d us\n", int(print_time.count()));
+```
+
+This example relies on the destruction of the timer to unlock the clock. If using a timer that doesn't get destroyed, you must remember to stop it manually to allow deep sleep to be entered again.
+
+**Recording high-resolution time points**
+```C++
+HighResClock::lock(); // take lock to ensure clock is steady (by preventing deep sleep)
+auto t1 = HighResClock::now();
+do_operation1();
+auto t2 = HighResClock::now();
+do_operation2();
+auto t3 = HighResClock::now();
+HighResClock::unlock();
+microseconds total = t3 - t1;
+microseconds stage1 = t2 - t1;
+microseconds stage2 = t3 - t2; 
+printf("total time   = %d us\n", int(total.count()));
+printf("stage 1 time = %d us\n", int(stage1.count()));
+printf("stage 2 time = %d us\n", int(stage2.count()));
+```
+
+**Blinking an LED at 1kHz**
+```C++
+Ticker ticker;
+ticker.attach([] { LED1 = !LED1; }, 500us);
+``` 
+
+### Using the real-time clock
+
+**Logging times**
+```C++
+auto woken = RealTimeClock::now();
+time_t t = RealTimeClock::to_time_t(woken);
+printf("Woken at %s", ctime(&t));
+do_work();
+t = RealTimeClock::to_time_t(woken + 3h);
+printf("Sleeping until %s", ctime(&t));
+ThisThread::sleep_for(3h);
+```
+
+Note that RTOS APIs currently only accept RTOS time points, which is why the above example uses `sleep_for`; you would get a compile error for `ThisThread::sleep_until(woken + 3h)`. If the work took a long time, then the schedule would drift away from a regular 3 hour period. See next example for an absolute-time-based approach.
+
+**Making helpers to combine clocks**
+
+The above example could be adapted to use absolute wake times with a helper function, guaranteeing a regular 3-hour period, regardless of how long the work takes:
+
+```C++
+void sleep_until_rtc(RealTimeClock::time_point abs_time)
+{
+    auto rel_time = abs_time - RealTimeClock::now();
+    ThisThread::sleep_until(Kernel::Clock::now() + rel_time);
+}
+
+auto woken = RealTimeClock::now();
+time_t t = RealTimeClock::to_time_t(woken);
+printf("Woken at %s", ctime(&t));
+do_work();
+auto wake = woken + 3h;
+t = RealTimeClock::to_time_t(wake);
+printf("Sleeping until %s", ctime(&t));
+sleep_until_rtc(wake);
+```
+
+Note that `sleep_until_rtc` would not respond to any adjustments made to `RealTimeClock` made while sleeping.
+
+### Being more generic
+
+**Allowing choice of clock**
+
+As the high-resolution and low-power clocks provide the same API, code can be written generically to work with either clock. For example:
```

Comment:
some example have a line break before the code block, some don't.

when converting to commonmark with pandoc, a new line is added after `...example:`.